### PR TITLE
improvement: schema validation error quality (phase 2)

### LINKED ISSUE #1394: Show a better error when paths are not formatted correctly
In the editor, we currently show this error message when a path does not start with a `/`:

image

This can be reproduced by any simple definition, like such:

image

Since we know exactly what is wrong in this case, it would be much better for end users if we told them that the path didn't follow the expected structure, which begins with a `/`.

### LINKED ISSUE #1480: OAS 3.0 securityScheme type "oauth2" not supported
In the components, if I put a "securityScheme" with type "oauth2", I get errors :

```
Schema error at components.securitySchemes['apiOAuth']
should NOT have additional properties
additionalProperty: flows

Schema error at components.securitySchemes['apiOAuth'].type
should be equal to one of the allowed values
allowedValues: apiKey, http, openIdConnect

Schema error at components.securitySchemes['apiOAuth'].flows.implicit
should NOT have additional properties
additionalProperty: tokenUrl
```

### Demonstration API definition

Based on the petstore example.

```yaml
openapi: "3.0.0"
info:
  version: 1.0.0
  title: Swagger Petstore
  license:
    name: MIT
servers:
  - url: http://petstore.swagger.io/v1
paths:
  /pets:
    get:
      summary: List all pets
      operationId: listPets
      tags:
        - pets
      parameters:
        - name: limit
          in: query
          description: How many items to return at one time (max 100)
          required: false
          schema:
            type: integer
            format: int32
      responses:
        '200':
          description: An paged array of pets
          headers:
            x-next:
              description: A link to the next page of responses
              schema:
                type: string
          content:
            application/json:    
              schema:
                $ref: "#/components/schemas/Pets"
        default:
          description: unexpected error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Error"
    post:
      summary: Create a pet
      operationId: createPets
      tags:
        - pets
      responses:
        '201':
          description: Null response
        default:
          description: unexpected error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Error"
  /pets/{petId}:
    get:
      summary: Info for a specific pet
      operationId: showPetById
      tags:
        - pets
      parameters:
        - name: petId
          in: path
          required: true
          description: The id of the pet to retrieve
          schema:
            type: string
      responses:
        '200':
          description: Expected response to a valid request
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Pets"
        default:
          description: unexpected error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Error"
components:
  schemas:
    Pet:
      required:
        - id
        - name
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
        tag:
          type: string
    Pets:
      type: array
      items:
        $ref: "#/components/schemas/Pet"
    Error:
      required:
        - code
        - message
      properties:
        code:
          type: integer
          format: int32
        message:
          type: string
  securitySchemes:
    apiOAuth:
      type: oauth2
      flows:
        implicit:
          authorizationUrl: 'https://myapi.com/oauth/authorize'
          tokenUrl: 'https://myapi.com/oauth/token'
          refreshUrl: 'https://myapi.com/oauth/token/refresh'
          scopes:
            -'write:pets': "modify pets in your account"

```

### Current Behavior
The "oauth2" type for securitySchemes is not implemented ?

### LINKED ISSUE #1489: Validation message when oneOf fails for example(s)
| Q                                   | A
| ----------------------------------- | -------
| Bug or feature request?             |   ¯\\_(ツ)_/¯
| Which Swagger/OpenAPI version?      |  OAS3
| Which Swagger-Editor version?       | 3.1.6
| How did you install Swagger-Editor? | -
| Which broswer & version?            | Chrome
| Which operating system?             |  Windows

### Demonstration API definition

```yaml
openapi: 3.0.0
info:
  version: "validation"
  title: Spec with both example and examples
paths:
  /myPath:
    get:
      responses:
        '200':
          description: OK
          content:
            text/plain:
              example: sample response
              examples:
                foo:
                  value: anonther response
```

### Expected Behavior
One error message that states that I can't have both `example` and `examples` in `paths['/myPath'].get.responses['200'].content['text/plain']`

### Current Behavior
Three error messages there neither one of them nails the error description.

### LINKED ISSUE #1511: Validation Parameters, vague errors that have regressed ( I believe )
| Q                                   | A
| ----------------------------------- | -------
| Bug or feature request?             |  Bug ( regression )
| Which Swagger/OpenAPI version?      | 2.0
| Which Swagger-Editor version?       | swaggerEditor/Ui: "3.1.9/g95b3f38a-dirty" / version: "3.2.2" / gitRevision: "g774a066"
| How did you install Swagger-Editor? |  Custom + Using Online version
| Which broswer & version?            | Chrome/59
| Which operating system?             | Linux Mint 18

### Demonstration API definition

```yaml
swagger: '2.0'
info:
  version: '1'
  title: oneOf validation fail
  description: ok

paths:
  /models/{name}:
    get:
      summary: Extra `example` field in parameter
      description: ok
      parameters: 
      - in: path
        name: name
        type: string
        required: true
        example: ok
      responses:
        200:
          description: Ok
```

### Expected Behavior
Validation should show that the "example" key is not allowed in "path" parameters.

### Current Behavior
Two validation errors occur...
```Schema error at paths['/models/{name}'].get.parameters[0]
should NOT have additional properties
additionalProperty: type, example, in, name, required
Jump to line 13
Schema error at paths['/models/{name}'].get.parameters[0].in
should be equal to one of the allowed values
allowedValues: body, header, formData, query
Jump to line 13
```

### Possible Solution
Hack the structural validation lib, to better handle oneOf cases?
Ie: `in: path` then enforce validation for path items.
Or catch the strings ( this was done before and has regressed ) and transform the error message.
And perhaps lastly, add custom validation to the cases where oneOf appear, and remove the bad structural messages.
### Context ( as a user )
Frustration on how to fix the spec, given the vague somewhat incorrect validation messages.

### LINKED ISSUE #1519: Online editor reports several nonsensical errors
Bug

| Q                                   | A
| ----------------------------------- | -------
| Bug or feature request?             |  Bug
| Which Swagger/OpenAPI version?      |3.0.0
| Which Swagger-Editor version?       |3.1.9
| How did you install Swagger-Editor? | npm i -S swagger-editor-dist
| Which broswer & version?            |chrome 62
| Which operating system?             | ubuntu 17.04

### Demonstration API definition
```yaml
openapi: 3.0.0
info:
  title: Example API
  description: Service API.
  version: 1.0.0
servers:
  - url: 'http://api.example.com'
paths:
  /login/local:
    post:
      tags:
        - login
      summary: Login using email or nickname
      description: Returns customer info and jwt access and refresh tokens
      security:
        - apiKey: []
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties:
                emailOrNickname:
                  type: string
                  required: true
                password:
                  type: string
                  required: true
      responses:
        '200':
          description: Login successful
          content:
            application/json:
              schema:
                type: object
                properties:
                  customer:
                    $ref: "#/components/schemas/Admin"
                  refreshToken:
                    type: string
        "401":
          description: "Unauthorized"
components:
  securitySchemes:
    apiKey:
      type: apiKey
      in: header
      name: x-application-key
  schemas:
    Admin:
      type: object
      properties:
        _id:
          type: string
        name:
          type: object
          properties:
            first:
              type: string
            last:
              type: string
        email:
          type: object
          properties:
            address:
              type: string
            verified:
              type: boolean
            token:
              type: string
        password:
          type: string
        recoveryToken:
          type: string
        lastLogin:
          type: string
          format: date-time
```

### Configuration (browser query string, constructor, config.yaml)

I'm using the default config

```js
const editor = SwaggerEditorBundle({
      dom_id: '#swagger-editor',
      layout: 'StandaloneLayout',
      presets: [
        SwaggerEditorStandalonePreset
      ]
    })
```

### Expected Behavior
I think my path definition is correct. I've read and re-read the 3.0 spec several times, I based my code on the examples on spec but I still get several weird errors: Mainly I don't want to use a component `$ref` on the `requestBody`. I've seen example code on the spec documentation and it looks like I should have that option

### Current Behavior
```
Errors
Hide
Schema error at paths['/login/local'].post.requestBody
should have required property '$ref'
missingProperty: $ref
Jump to line 17
Schema error at paths['/login/local'].post.requestBody
should match exactly one schema in oneOf
Jump to line 17
Schema error at paths['/login/local'].post.requestBody.content['application/json']
should have required property 'examples'
missingProperty: examples
Jump to line 19
Schema error at paths['/login/local'].post.requestBody.content['application/json']
should match exactly one schema in oneOf
Jump to line 19
Schema error at paths['/login/local'].post.requestBody.content['application/json'].schema
should have required property '$ref'
missingProperty: $ref
Jump to line 20
Schema error at paths['/login/local'].post.requestBody.content['application/json'].schema
should match exactly one schema in oneOf
Jump to line 20
Schema error at paths['/login/local'].post.requestBody.content['application/json'].schema.properties['emailOrNickname']
should have required property '$ref'
missingProperty: $ref
Jump to line 23
Schema error at paths['/login/local'].post.requestBody.content['application/json'].schema.properties['emailOrNickname']
should match exactly one schema in oneOf
Jump to line 23
Schema error at paths['/login/local'].post.requestBody.content['application/json'].schema.properties['emailOrNickname'].required
should be array
Jump to line 25
Schema error at paths['/login/local'].post.requestBody.content['application/json'].schema.properties['password']
should have required property '$ref'
missingProperty: $ref
Jump to line 26
Schema error at paths['/login/local'].post.requestBody.content['application/json'].schema.properties['password']
should match exactly one schema in oneOf
Jump to line 26
Schema error at paths['/login/local'].post.requestBody.content['application/json'].schema.properties['password'].required
should be array
Jump to line 28
```

### Context
I'm trying to migrate some bad documentation to something better using swagger 3

### LINKED ISSUE #1672: OAS 3.0.1: apiKey security scheme has incomprehensible validation errors
| Q                                   | A
| ----------------------------------- | -------
| Bug or feature request?             |  Bug
| Which Swagger/OpenAPI version?      | 3.0.1
| Which Swagger-Editor version?       | 3.2.9
| How did you install Swagger-Editor? | yarn
| Which broswer & version?            | Chrome latest
| Which operating system?             | Fedora 27

### Demonstration API definition
```yaml
---
openapi: 3.0.1
info:
  title: Test
  version: 1.0.0
paths: {}
components:
  schemas: {}
  securitySchemes:
    headerFoo:
      type: apiKey
      in: header
  links: {}
  callbacks: {}

```

### Configuration (browser query string, constructor, config.yaml)
Completely stock

### Expected Behavior
I expected it to squawk that `headerFoo` lacked a `name` parameter.

### Current Behavior
```
Schema error at components.securitySchemes['headerFoo']
should NOT have additional properties
additionalProperty: in

Schema error at components.securitySchemes['headerFoo'].type
should be equal to one of the allowed values
allowedValues: http, oauth2, openIdConnect
```

This is weird, because it explicitly _enumerates_ the valid types for this...and omits `apiKey`. If I fix the scheme def, it's fine:

```yaml
    headerFoo:
      type: apiKey
      name: bob
      in: header
```

### Context
I'm writing an OpenAPI-first web framework for Ruby/Rack, called [Modern](https://github.com/modern-project/modern-ruby). As I build my OpenAPI spec generator, I keep occasionally testing against the editor (because [the only Ruby OAPI gem I can find](https://github.com/kevindew/openapi3_parser) is incomplete and I haven't extended my test pipeline to automate the use of swagger-js) and this one made me go "buh?".

### LINKED ISSUE #1709: Relative URL for openIdConnectUrl has errors
| Q                                   | A
| ----------------------------------- | -------
| Bug or feature request?             |  Bug
| Which Swagger/OpenAPI version?      | 3.0.0
| Which Swagger-Editor version?       | 3.4.1
| How did you install Swagger-Editor? | npm, and hosted swagger.io
| Which browser & version?            | chrome 64.0.3282.186 
| Which operating system?             |  Windows 10 64-bit

### Demonstration API definition

Sample [Gist](https://gist.githubusercontent.com/jblazek/5b7cad3e3ec8002c3bf1a6e80fcc8408/raw/deaca3e01fd401d34dc00a1ababa4ee773463c61/openid-relative-url.yaml)

Using swagger-editor, I'm getting errors when using the example code from [openid](https://swagger.io/docs/specification/authentication/openid-connect-discovery/) 

```yaml
servers:
  - url: https://api.example.com/v2

components:
  securitySchemes:
    openId:
      type: openIdConnect
      openIdConnectUrl: /.well-known/openid-configuration
```

If I change the `openIdConnectUrl` to be an absolute URL, I don't get issues. 

### Expected Behavior

I would expect the relative URL to be accepted, since the documentation and example say it's allowed.

### Current Behavior

Currently, several errors are displayed

```
Schema error at components.securitySchemes['openId']
should NOT have additional properties
additionalProperty: openIdConnectUrl
Jump to line 25
Schema error at components.securitySchemes['openId'].type
should be equal to one of the allowed values
allowedValues: apiKey, http, oauth2
Jump to line 26
Schema error at components.securitySchemes['openId'].openIdConnectUrl
should match format "url"
format: url
Jump to line 27
```

### Context

I want to use openid connect, but don't want to specify the full URL, since it depends on the servers I have configured.

### LINKED ISSUE #1711: Swagger 2.0 's security scheme validated against OAS 3.0 instead of swagger 2.0
_From @pbarbier on March 23, 2018 18:15_

When editing a swagger 2.0 yaml content featuring oauth2 security schemes like below:

```
swagger: "2.0"

info:
  title: wow
  version: 1.0.0
  
paths:
  /:
    get:
      description: wow
      responses:
        200:
          description: ok

securityDefinitions:
  oauth2_Password:
    type: oauth2
    tokenUrl: /authorizationserver/oauth/token
    flow: password
    scopes:
      basic: ''
  oauth2_client_credentials:
    type: oauth2
    tokenUrl: /authorizationserver/oauth/token
    flow: application
    scopes:
      extended: ''
  apikey:
    type: apiKey
    name: myAuth
  basic:
    type: basic
    in: header
```
Several nonsensical errors are reported:

```
Structural error at securityDefinitions.oauth2_Password
should NOT have additional properties
additionalProperty: tokenUrl, flow, scopes

Structural error at securityDefinitions.oauth2_Password.type
should be equal to one of the allowed values
allowedValues: basic, apiKey

Structural error at securityDefinitions.oauth2_Password.tokenUrl
should match format "uri"
format: uri

Structural error at securityDefinitions.oauth2_Password.flow
should be equal to one of the allowed values
allowedValues: implicit, application, accessCode

Structural error at securityDefinitions.oauth2_client_credentials
should NOT have additional properties
additionalProperty: tokenUrl, flow, scopes

Structural error at securityDefinitions.oauth2_client_credentials.type
should be equal to one of the allowed values
allowedValues: basic, apiKey

Structural error at securityDefinitions.oauth2_client_credentials.tokenUrl
should match format "uri"
format: uri

Structural error at securityDefinitions.oauth2_client_credentials.flow
should be equal to one of the allowed values
allowedValues: implicit, password, accessCode

Structural error at securityDefinitions.apikey
should NOT have additional properties
additionalProperty: name

Structural error at securityDefinitions.apikey.type
should be equal to one of the allowed values
allowedValues: basic, oauth2

Structural error at securityDefinitions.basic
should NOT have additional properties
additionalProperty: in

Structural error at securityDefinitions.basic.type
should be equal to one of the allowed values
allowedValues: apiKey, oauth2

```

| Q                               | A
| ------------------------------- | -------
| Bug or feature request?         |  Bug
| Which Swagger/OpenAPI version?  |  Swagger 2.0
| Which Swagger-UI version?       |  3.13.0
| How did you install Swagger-UI? |  Use through editor.swagger.io
| Which browser & version?        | Chrome latest
| Which operating system?         | WIN10

### Demonstration API definition
See above

### Configuration (browser query string, constructor, config.yaml)
regular use of editor.swagger.io

### Expected Behavior
No error should be shown for this part of the yaml

### Current Behavior
See above

### Context
The error messages are misleading. They claim the content is not valid when it is.

_Copied from original issue: swagger-api/swagger-ui#4364_

### LINKED ISSUE #1797: Within an OperationObject, a ParametersObject doesn't accept in: path as a valid location
### Q&A (please complete the following information)
 - Method of installation: Public hosted Testing
 - Swagger-Editor version: Public hosted @ https://editor.swagger.io
 - Swagger/OpenAPI version: OpenAPI 3.0

### Content & configuration
Example Swagger/OpenAPI definition:
```yaml
openapi: 3.0.1
info:
  title: Foo OpenApi Spec
  version: 1.0.0
servers:
  - url: 'https://localhost/BasePath'
paths:
  '/Account/{id}':
    get:
      description: |-
      operationId: Account1
      parameters:
        - name: id
          in: path
          allowEmptyValue: true
          schema:
            type: integer
            format: int32
      responses:
        '200':
          description: |-
```

Swagger-Editor configuration options:
n/a - using editor.swagger.io as a validator

### Describe the bug you're encountering
Validation error of OpenApi schema - within an OperationObject, a ParametersObject doesn't accept in: **path** as a valid location

### To reproduce...
1. Paste the above YAML into the editor on editor.swagger.io
2. Read the Errors on the right side

### Expected behavior
No validation Errors and for "path" to be an acceptable "in" location

### Screenshots

image

### LINKED ISSUE #1808: Confusing error message "should NOT be valid" caused by empty responses
### Q&A (please complete the following information)
 - OS: Windows
 - Browser: Chrome 67
 - Method of installation: http://editor.swagger.io
 - Swagger-Editor version: 3.6.0
 - Swagger/OpenAPI version: 2.0, 3.0

### Example Swagger/OpenAPI definition
2.0:
```yaml
swagger: "2.0"
info:
  version: 0.0.0
  title: test
paths:
  /foo:
    get:
      responses: {}
```
3.0:
```yaml
openapi: 3.0.1
info:
  version: 0.0.0
  title: test
paths:
  /foo:
    get:
      responses: {}
```

### Describe the bug you're encountering
The `response: {}` line causes two validation errors:

1. should NOT have less than 1 properties
2. should NOT be valid

The second error does not make sense.

### Expected behavior
No "should NOT be valid" error.

### Screenshots
error

### LINKED ISSUE #1832: "should be an array" error caused by misspelled "type" values
### Q&A (please complete the following information)
 - OS: Windows 7
 - Browser: Chrome 67
 - Method of installation: http://editor.swagger.io
 - Swagger-Editor version: 3.6.3
 - Swagger/OpenAPI version: 2.0

### Example Swagger/OpenAPI definition:
```yaml
swagger: '2.0'
info:
  title: test
  version: 0.0.0
paths: {}

definitions:
  Model1:
    type: int
```

### Describe the bug you're encountering
A misspelled `type` value triggers an odd error

> should be array

### Expected behavior
No "should be array" error.

### Screenshots
image

### LINKED ISSUE #1853: Misleading error message claiming allowed values in "in" of a parameter only "query, header, cookie" without "path"
### Q&A (please complete the following information)
 - OS: Linux
 - Browser: Firefox
 - Version: 61.0.1
 - Method of installation: docker
 - Swagger-Editor version: swaggerapi/swagger-editor docker image,latest,image ID 9fb9dba980ed
 - Swagger/OpenAPI version: OpenAPI 3.0

### Content & configuration
image

Example Swagger/OpenAPI definition:
```yaml
openapi: 3.0.1
info:
  title: Contact List API
  description: CRUD a simple Contact item.
  version: '0.1'
  termsOfService: ''
  contact:
    name: Bing Ren
    email: bingtimren@gmail.com
  license:
    name: Free to use
    url: ''
servers:
  - url: 'http://localhost:8080/contactList/0.1/'
    description: SAM local api
paths:
  /contact:
    get:
      summary: Get contacts as a list
      description: Get a list of contacts
      operationId: getContactAsList
      responses:
        '200':
          description: Successful response
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Contact'
        '404':
          description: Not found response
          content:
            text/plain:
              schema:
                title: Weather not found
                type: string
                example: Not found
  /contact/{contactId}:
    get:
      summary: Get a single contact by Id
      operationId: getContactById
      parameters: 
        - name: contactId
          in: path
          description: ID of contact to return
          required: true
          # schema:
            # type: integer
            # format: int64
security:
  - app_id: []
components:
  schemas:
    ContactList:
      title: Contact List
      type: array
      items:
        $ref: '#/components/schemas/Contact'
    Contact:
      title: Contact Item
      type: object
      required:
        - id
      properties:
        id:
          type: integer
          description: Internal parameter
          format: int32
          example: 8166
        name:
          type: string
          description: Contact's full name
          example: Winston Smith
        email:
          type: string
          description: Email
          example: winston.smith@gmail.com
        phone:
          type: string
          description: Phone number
          example: +1 999999999
          pattern: '\+[0-9 ]+'
        age:
          type: integer
          description: Age
          format: int32
          example: 40
          minimum: 0
          maximum: 150
  securitySchemes:
    app_id:
      type: apiKey
      description: >-
        API key to authorize requests. If you don't have an OpenWeatherMap API
        key, use `fd4698c940c6d1da602a70ac34f0b147`.
      name: appid
      in: query

```

Swagger-Editor configuration options:
```
I didn't configure the editor, just launched the docker image
```

### Describe the bug you're encountering

This is a file I'm working on. When I still have not defined the schema for a path parameter, editor reports an error at the "in" property, claiming the allowed values are only "query, header, cookie". This is very confusing and in fact wrong. If I just complete the parameter definition by providing it's schema, this error is not reported, although the value for "in" is still "path".

As a learner myself to the OpenAPI specification I was quite confused by the misleading error, and spent much time consulting with documents and examples, then finally concluded that it's a false alarm.

```
Schema error at paths['/contact/{contactId}'].get.parameters[0].in
should be equal to one of the allowed values
allowedValues: query, header, cookie
Jump to line 43
```

### To reproduce...

Load the sample definition

### Expected behavior

The error should not be reported in the first place. Or, if that's not possible or difficult, at least make the message not so misleading. 

### Screenshots

See above

### PR DESCRIPTION
**Pull request by shockey**
### Description

This PR is the second phase of my schema validation improvement work (phase 1 was introduced via #1982 and #1984).

It includes changes to the OpenAPI validation schema documents, our AJV instance, and one small modification to our local JSON Schema meta-document.

A few techniques have been deployed here:
- migrating usage of `oneOf` to artificial `switch` constructs that pivot the validation schema based on specific keys
  - most helpfully, using inline content schemas unless a `$ref` property is present
- providing custom, contextual error messages for assertion keywords within specific Objects
- flattening unnecessary constructs in the validation schemas

Notably, this modifies generated errors to report their locations with JSON Pointer-compliant strings. Downstream consumers of schema errors may need to make changes accordingly, but this is not a breaking change since schema errors are within an internal subsystem.

### Motivation and Context

Fixes #1853.
Fixes #1832.
Fixes #1808.
Fixes #1797.
Fixes #1711.
Fixes #1709.
Fixes #1672.
Fixes #1519.
Fixes #1511.
Fixes #1489.
Fixes #1480.
Fixes #1394.

### How Has This Been Tested?

No tests for now - I may add tests for the linked issues before merging.

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [x] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [ ] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by shockey** [comment]
please build

File: src/plugins/json-schema-validator/oas3-schema.yaml [hunk 1]
```diff
@@ -597,38 +639,48 @@ definitions:
...
+            required:
+            - example
+        - not:
+            required:
+            - examples
+    - if: 
+        not:
+          required: [schema, content]
+      then: false
+    errorMessage: "should have either a `schema` or `content` property"
```
Comment (shockey) [inline src/plugins/json-schema-validator/oas3-schema.yaml]: does this make the content `not`s confusing?

File: src/plugins/json-schema-validator/oas3-schema.yaml [hunk 2]
```diff
@@ -748,11 +796,37 @@ definitions:
...
+        properties:
+          type:
+            enum: [openIdConnect]
+      then:
+        $ref: "#/definitions/OpenIdConnectSecurityScheme"
+    - then:
+        properties:
+          type:
+            type: string
+            enum: [apiKey, http, oauth2, openIdConnect]
```
Comment (shockey) [inline src/plugins/json-schema-validator/oas3-schema.yaml]: should `properties` be required?

File: src/plugins/json-schema-validator/swagger2-schema.yaml [hunk 3]
```diff
@@ -850,13 +880,30 @@ definitions:
...
+            then: { $ref: "#/definitions/oauth2ApplicationSecurity" }
+          - if: { properties: { flow: { enum: [accessCode] } } }
+            then: { $ref: "#/definitions/oauth2AccessCodeSecurity" }
+          - then: 
+              properties:
+                flow:
+                  enum: [implicit, password, application, accessCode]
+      - then: 
+          properties:
+            type:
```
Comment (shockey) [inline src/plugins/json-schema-validator/swagger2-schema.yaml]: should `type` be required?

File: test/unit/plugins/json-schema-validator/test-documents/1672.yaml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+meta:
+  title: "#1672 - Header Security Scheme Object lacking a `name`"
+input:
+  openapi: 3.0.1
```
Comment (shockey) [inline test/unit/plugins/json-schema-validator/test-documents/1672.yaml]: should be `3.0.0`

File: test/unit/plugins/json-schema-validator/test-documents/1853.yaml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+meta:
+  title: "#1853 - OpenAPI 3.0 path parameter missing both `schema` and `content`"
+input:
+  openapi: 3.0.1
```
Comment (shockey) [inline test/unit/plugins/json-schema-validator/test-documents/1853.yaml]: should be 3.0.0

File: test/unit/plugins/json-schema-validator/test-documents/1711.yaml [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      message: should be an absolute URI
+      path: [securityDefinitions, oauth2_client_credentials, tokenUrl]
+      source: structural
+- name: apiKey definition missing `in` property
+  input:
+    swagger: "2.0"
+    info:
+      title: wow
+      version: 1.0.0
+    paths:
```
Comment (shockey) [inline test/unit/plugins/json-schema-validator/test-documents/1711.yaml]: change to `paths: {}`

File: test/unit/plugins/json-schema-validator/test-documents/1711.yaml [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        missingProperty: in
+      path: [securityDefinitions, apikey]
+      source: structural
+- name: basic definition with a superfluous `in` property
+  input:
+    swagger: "2.0"
+    info:
+      title: wow
+      version: 1.0.0
+    paths:
```
Comment (shockey) [inline test/unit/plugins/json-schema-validator/test-documents/1711.yaml]: change to `paths: {}`

**Comment by shockey** [comment]
Fin
